mmf/datasets/multi_dataset_loader.py

```diff
@@ -132,17 +134,28 @@ def build_datasets(self, config):
             self.datasets.append(dataset_instance)
 
             if hasattr(dataset_instance, "__len__"):
-                self._per_dataset_lengths.append(len(dataset_instance))
-                self._total_length += len(dataset_instance)
+                dataset_instance_length = len(dataset_instance)
+                if dataset_instance_length:
+                    self._per_dataset_lengths.append(dataset_instance_length)
+                    self._total_length += dataset_instance_length
+                else:
+                    # Filter out datasets with no items
+                    self.datasets.pop()
 
         self._num_datasets = len(self.datasets)
-        self.current_index = 0
-        self.current_dataset = self.datasets[self.current_index]
+        if self._num_datasets:
+            self.current_index = 0
+            self.current_dataset = self.datasets[self.current_index]
 
         self._infer_dataset_probabilities()
+        self._did_build_datasets = True
 
     def build_dataloaders(self):
-        assert len(self._datasets) > 0, "Call build_datasets first"
+        if not self._did_build_datasets:
+            assert len(self._datasets) > 0, "Call build_datasets first"
+
+        if len(self.datasets) == 0:
```

Comment:
I think this logic is getting unnecessarily complicated for a single use case. We should fail fast when there is no datasets, or use try except python style.